### ISSUE TITLE
[api-extractor] Support "import * as __ from __" for local paths

### ISSUE THREAD

**Issue by octogonz** [microsoft/rushstack#1029]
PR https://github.com/Microsoft/web-build-tools/pull/1002 adds support for declarations like this:

```ts
import * as React from 'react';
```

This is a trivial implementation, where we pretend that `*` is an actual name exported by the `"react"` package.  It does not work for the general form that introduces a new namespace, for example:

```ts
// For example, "Button" in controls.ts becomes "controls.Button"
import * as controls from './controls';
```

This form is called `ts.SyntaxKind.NamespaceImport` because it is essentially a more general form of the `namespace controls { }` declaration kind.

As such, it should probably produce an `ApiItemKind.Namespace` item.

**Comment by pgonzal** [comment]
@gf3 shared a branch https://github.com/gf3/api-extractor-issue/commit/30de59d08a1592222ae81d884948029d7ab08749 that repros this issue.

**Comment by pgonzal** [comment]
Some other examples were given in the comments for issue #663

**Comment by shlomiassaf** [comment]
@pgonzal I think that those imports should be handled differently, from a usage perspective and rather then semantic...

```ts
import * as React from 'react';
import * as controls from './controls';

export class ABC {
  p1: React.Component;
  p2: controls.Control;
}
```

The code above is similar to this:

```ts
import { Component }  from 'react';
import { Control } from './controls';

export class ABC {
  p1: Component;
  p2: Control;
}
```

The local import is purely semantic...
When generating docs it doesn't make sense to render `p2: controls.Control;` so the library should find the direct reference and ignore the type.

When importing from third party, it's a different story and the way it is rendered is up to the user, just need to make sure there's enough info...

**Comment by shlomiassaf** [comment]
@octogonz Thanks for clear explanation.

I think that the underlaying issue is more about the concept and direction of this project.
This tool is used for several tasks, document generation is one of them.

We can agree that generating documentation from the `json` output can happen is a lot of ways, so many that I can't count :)

At it's current state, the project does provide a way to change the behaviour, configuring it, etc...

Taking a lot of assumptions is risky because it means that it will only do and support `things that occur in your code bases` and that's not a very good approach for a public project.

I see great importance in `tsdoc` and this tool and I think that wide adoption will help the community because there are a lot of tools for this job, each with pros/cons and having a unified tool from MS is probably the best solution.

But, for that we need flexibility, understanding that your code base might not be the same as another's code base.

For example, let's say I built an angular library that uses a lot of symbols from various `@angular` libraries and also symbols from `rxjs`

Now of course all of them are not exported, but in my doc generator I can decide to generate links for them, based on some schema... links to the external documentation for these libraries...

So if a new `ApiItem` is created hosting the (not) exported symbol with it's package name, the doc generator can do the magic...

In angular it's a very simple schema: `https://angular.io/api/[PACKAGE NAME]/[SYMBOL NAME]`
Examples:

```ts
import { HttpHeaders } from '@angular/common/http';
```
https://angular.io/api/common/http/HttpHeaders


```ts
import { Component } from '@angular/core';
```
https://angular.io/api/core/Component

Usually, since documentation is auto-generated, a schema always exists...

**Comment by jeremymeng** [comment]
We are currently using the following pattern. Is there a recommended workaround?

```typescript
import * as Models from "../lib/generated/lib/models";
...
export { Models };
```

**Comment by gf3** [comment]
is there anything we can do to help move this along? we're hoping to use this tool at Shopify to build out some of our public facing documentation for a few libraries and tools

**Comment by mterrel** [comment]
@octogonz After looking at all the similar tools out there for TypeScript projects, this definitely seems like the right choice. But we do use the same pattern as @jeremymeng, so we're blocked by this issue.

I'd rather pitch in and help out here than spend time setting up a different tool, only to come back to api-extractor later. So I'm definitely willing to put in some work to fix this. I've got quite a bit of experience working with the TS compiler API and the AST, but am new to api-extractor. Reading through the previous comments, it sounds like the fix may be a bit involved, so I thought I'd reach out first to see if 1) you'd like some help here and 2) if you've got any high level guidance that you'd like me to follow. I'm happy to jump on Slack or a call or whatever is easiest if you want to chat.

**Comment by octogonz** [comment]
@mterrel That would be great and much appreciated!  Ping me [on Gitter](https://gitter.im/web-build-tools/web-build-tools) and we can figure something out. Very exciting!

**Comment by xirzec** [comment]
Any chance this is still being worked on? It would be lovely to have for the Azure SDK.

**Comment by adventure-yunfei** [comment]
Blocking upgrading from AE6 by this, too...

The syntax `export * as something from "./localfile` works on older api-extractor version (tested on 6.1.3), but fails on latest version (7.4.2).

I remember that I should have fixed this case by https://github.com/microsoft/web-build-tools/pull/773.  Is this a regression caused by AE7 refactor?

**Comment by octogonz** [comment]
> Is this a regression caused by AE7 refactor?

@adventure-yunfei  Yes, in AE6 the .api.ts report was created using the original analysis engine, whereas the .d.ts rollups were created using the newer (more accurate) analysis engine.  Thus, in AE6 the `export * as something from "./localfile"` worked for the API report, but it was broken when generating .d.ts rollups.  (The [example](https://github.com/adventure-yunfei/api-documenter-example-case/blob/master/api-extractor.json) that you shared does not encounter this problem because .d.ts rollups are disabled.)

In AE7 everything, now uses the newer engine. The older engine is completely removed.  That is why this problem has resurfaced.  Fixing it is somewhat more difficult, because the solution needs to produce a correct .d.ts rollup.

I don't have time to work on this myself right now (I'm about to start work on some new Rush features), but if someone wants to volunteer, I'd be happy to help with the design.

Here's some initial thoughts.

### Interesting edge cases

We need to enumerate the weird edge cases, and come up with a design for how we would represent them in the .d.ts rollup.  For example, consider this pathological case:

```ts
import * as ns1 from './File1';
import * as ns2 from './File1';
import { thing } from './File1';

export { ns1 };
export { ns2 };
export { thing };
```

What happens to the string `./File1` when we rolls that up into a single rollup.d.ts file?

It's perfectly okay for our answer to be that "API Extractor supports `import *` for certain common usages, but certain other usages are NOT supported because (1) they cannot be rolled up reasonably and/or (2) they can be easily rewritten to better forms." It's okay to impose constraints. However, if we do that, it /is/ important to be able to clearly state what is and isn't supported. The rule can't be vague.

This exercise will also lead us to understand how to model `ns1`.  Is it a conventional `namespace` (i.e. [AstNamespace](https://github.com/microsoft/web-build-tools/blob/036f4dce6f7f5791a391c99495664d12d3b95e00/apps/api-extractor-model/src/model/ApiNamespace.ts#L44)?  Or is it some new construct?  (These might be easy answers BTW; I haven't had time to think about it in any depth.)

### Getting started

With the new analyzer, the starting point is in the [_tryMatchImportDeclaration()](https://github.com/microsoft/web-build-tools/blob/c5e13a01bf88d5c08def2db5b46f27df97083ac7/apps/api-extractor/src/analyzer/ExportAnalyzer.ts#L413) and [_tryMatchExportDeclaration()](https://github.com/microsoft/web-build-tools/blob/c5e13a01bf88d5c08def2db5b46f27df97083ac7/apps/api-extractor/src/analyzer/ExportAnalyzer.ts#L337) methods, which parse the `import` and  `export` statements and decide what to do with them.

The website has some [technical notes](https://api-extractor.com/pages/contributing/architecture/) about the new architecture that might he helpful, along with debugging instructions.  

If someone wants to pursue this, ping me on Gitter and I can provide more help.

**Comment by adventure-yunfei** [comment]
After some investigation, I think `namespace` should be the way. (although `namespace` should actually be a deprecated feature compared to (external) `module`).

### Syntax for module import

There're two ways to emit .d.ts rollup file for source file:
```ts
// foo.ts
import * as bar from './bar';
export { foo }
```

1. Use exactly the (external) `module` syntax:
    ```ts
    // bundle.d.ts
    declare module "<internal-key>/bar" {
      // ...
    }
    import * as bar from "<internal-key>/bar";
    export { bar }
    ```
    - pros: keep the same ts syntax node with source file
    - cons: pollute global module declaration
2. Use similar `namespace` syntax:
    ```ts
    // bundle.d.ts
    declare namespace bar {
        // ...
    }
    export { bar }
    ```
    - pros: keep export/global declaration clean
    - cons: `namespace` is actually not `module`

From typescript language, `namespace` (previously internal module) and `module` (previously external module) works in very similar way. The second way should be better.

### Implementation Step

For .d.ts rollup, there are two phases: analyze (to be included in api-extractor Collector Entities) and output (to *.d.ts bundle).

#### Analyze Phase

Right now I've made a prototype to first support module import for analyzer (https://github.com/adventure-yunfei/web-build-tools/pull/2), and made it work for api json generator (for some demo), by:
- creating a new `AstImportInternal` entity. why creating a new one:
  - it should not be another `AstSymbol`. `AstSymbol` implies some special logic, e.g. always referenced from outside by root symbol (so always analyze from the top, and only put the top symbol to collector entities))
  - it may be another `AstImport`, but currently `AstImport` only indicates **external** import. So either creating a new kind of entity, or extending `AstImport` definition.
- analyze it like `AstSymbol` (related declarations, file reference, etc), just like a normal local declaration
- process it in api generator like namespace.

#### Output Phase

When trying to actually make dts rollup work for module import, I've found many, many edge cases to be handled. So before implementation, I'd first share my syntax propose here.

##### *.d.ts Rollup syntax propose by namespace:

The idea is:
- If no module import, generate the same output as before (will not create unnecessary namespace wrapper)
- If has module import, generate top namespace declaration for each module import. Other standalone declaration for the module content should be merged to that namespace.
- Won't support default export inside module support (cannot find a way to support it in namespace declaration...)

**a)** With `input-a.ts`:
```ts
export class Foo {}
```

1. and `input-index.ts`:
    ```ts
    import * as a from './input-a.ts';
    export { a };
    ```
    will output:
    ```ts
    declare namespace a {
      export class Foo {}
    }
    export { a };
    ```
2. and `input-index.ts`:
    ```ts
    import * as a from './input-a.ts';
    export { Foo } from './input-a.ts';
    ```
    will output (the same as before):
    ```ts
    declare class Foo {}
    export { Foo }
    ```
3. and `input-index.ts` (compared to case 2):
    ```ts
    import * as a from './input-a.ts';
    export { a }
    export { Foo } from './input-a.ts';
    ```
    will output:
    ```ts
    declare class namespace a {
      export class Foo {}
    }
    import Foo = a.Foo;
    export { Foo }
    ```

**b)** With `input-a.ts`:
```ts
export class Foo {}
```
and another `input-b.ts`:
```ts
import * as a from './input-a';
export { a }
```

1. and `input-index.ts`:
    ```ts
    import * as b from './input-b';
    import * as a from './input-a';
    export { a, b }
    ```
    will output:
    ```ts
    declare namespace a_0 {
      export class Foo {}
    }
    declare namespace b {
      import a = a_0;
      export { a };
    }
    export { a_0 as a }
    ```

**c)** With `input-a.ts`:
```ts
import * as React from 'react';
import { map } from 'lodash';
export { React, map }
```
1. and `input-index.ts`:
    ```ts
    import * as a from './input-a';
    export { a }
    ```
    will output:
    ```ts
    import * as React_0 from 'react';
    import * as lodash from 'lodash';
    declare namespace a {
      import React = React_0;
      import map = lodash.map;
      export { React, map };
    }
    export { a }
    ```

**d)** With `input-a.ts`:
```ts
export class Foo {}
export default class Bar {}
```

1. and `input-index.ts`:
    ```ts
    import * as a from './input-a';
    export { a }
    ```
    will outout (**do not support default import**):
    ```ts
    declare namespace a {
      export class { Foo }
    }
    export * as a;
    ```

See, the namespace syntax is very complex due to declaration limitation inside namespace.

Note, there seems to be some issues when processing namespace multi-export. (see https://github.com/microsoft/rushstack/issues/1600 found recently)

##### *.d.ts Rollup syntax propose by module:

Since the original declarations are (external) module, to generate output by module syntax, it's a straight way.

For a complex example, with `input-a.ts`:

```ts
import * as React from 'react';
import { map } from 'lodash';
export class Foo {}
```
and `input-b.ts`:
```ts
import * as a from './input-a';
export { a }
export default class Bar {}
```

1. and `input-index.ts`:
    ```ts
    import * as b from './input-b';
    export { b };
    export { Foo } from './input-a';
    ```
    will output:
    ```ts
    declare module "__internal__/input-a" {
      import * as React from 'react';
      import { map } from 'lodash';
      declare class Foo {}
      export { React, map, Foo }
    }

    declare module "__internal__/input-b" {
      import * as a from "__internal__/input-a";
      declare class Bar {}
      export { a }
      export default Bar;
    }

    import * as b from "__internal__/input-b";
    import { Foo } from "__internel__/input-a";
    export { b, Foo }
    ```

Also, only create module wrapper when required.

For implementation, module syntax should be better.

**Comment by rbuckton** [comment]
I'm running into this in esfx now. Is there a workaround (either to tell api-extractor to ignore the declaration, or otherwise)?

**Comment by rbuckton** [comment]
> ... (although `namespace` should actually be a deprecated feature compared to (external) `module`).

I would argue that `namespace` isn't deprecated, though users are often encouraged to use external modules. There are a number of valid uses for `namespace`.

My recommendation for `import * as x from "..."; export { x };` would be to define `x` as `namespace x { ... }`.

Also keep in mind that `export * as x from "..";` is now (Stage 4)[https://github.com/tc39/proposal-export-ns-from] and is likely to land in TS [very soon](https://github.com/microsoft/TypeScript/pull/34903).

**Comment by octogonz** [comment]
> I'm running into this in esfx now. Is there a workaround (either to tell api-extractor to ignore the 
> declaration, or otherwise)?

We would definitely accept a PR that identifies unsupported cases and omits them from the output. That way people could ignore the error if they are okay with it.

We could also consider a TSDoc tag such as `@ignore` or `@omit` that would direct API Extractor to skip a problematic declaration.

Longer term, we'd really like to implement this feature, since people ask about it a lot. 

> My recommendation for `import * as x from "..."; export { x };` would be to define `x` as `namespace x { ... }`

Could you give a more complete syntax example?

When I encountered this in the past, I tried to rewrite `import * as x` to be `namespace x`, but I wasn't able to figure out the syntax for it.

**Comment by rbuckton** [comment]
I'm currently working around this with a very ugly and hacky patch that wires up a custom `AstSymbol` and `AstDeclaration` for a synthetic `export declare namespace x { ... }` in `ExportAnalyzer.prototype._tryMatchImportDeclaration`. Its not something I would say is "production quality" though.

**Comment by rbuckton** [comment]
It basically converts something like this:
```ts
// a.d.ts
export declare function f() ...
export declare function g() ...

// index.d.ts
import * as a from "./a";
export { a };
```

Into this:
```ts
// index.d.ts
export declare namespace a {
  export function f() ...
  export function g() ..
}
```

Although it does this by creating `AstDeclartion` and `AstSymbol` objects and reparenting the exports of `a.d.ts`. For reference, I've attached to this comment the (very hacky) JavaScript shim I implemented for this behavior.

[patchApiExtractor.zip](https://github.com/microsoft/rushstack/files/3969922/patchApiExtractor.zip)

**Comment by adventure-yunfei** [comment]
I got little time to work on this case.
Mentioned previously, while the `namespace` proposal has simpler syntax, it cannot express:
1. `default` export
2. re-export from external package (e.g. `export { map } from 'lodash';`)

The `default` case seems to be a rare case (allowing user to: `import foo from 'foo';  Foo.default.someApi(); ` ). The `re-export` case may be reasonal.

Now I prefer the `module` syntax. Since it may take some effort to implement, I'd like to get some opinions about the choice.  @octogonz

**Comment by paztis** [comment]
for me the namespace usage is a workaround.
The export * as X from 'Y' is a standard usage and must be supported by api-extractor.
I will not change my code source to allow a tool to correctly work. It may be the invert. Tools need to fit to the code.

For now due to this case, the api extraction is crashing, and totally block the usage.

**Comment by mckn** [comment]
What is the status on this issue, is there anything I can do to help out?

**Comment by paztis** [comment]
For now it is still throwing an exception in the processing.
rbuckton solution can temporary unlock the problem but when testing it I encounter an error after in Collector.ts `AstDeclaration.declarationMetadata is not expected to have been initialized yet`
I don't really know how to prevent it.

An idea might be to preprocess the files to replace all the `import * as AliasedName from 'moduleName'` into
```
import {
  A as tmpA,
  B as tmpB,
  C as tmpC
} from 'moduleName';
declare namespace AliasedName {
  export {
    tmpA as A,
    tmpB as B,
    tmpC as C
  }
}
```
but this imply to process everything one time to be able to understand A, B and C are all the named exports of `moduleName`, or to inject it during the processing

Any idea how to do it ?

**Comment by adventure-yunfei** [comment]
> but this imply to process everything one time to be able to understand A, B and C are all the named exports of `moduleName`, or to inject it during the processing

To workaround by preprocessing, you can simply using typescript custom transformer feature. There seems no standard doc for ts tranformer, but it's already used by e.g. `ts-loader`. You can also search google to learn, e.g. https://levelup.gitconnected.com/writing-a-custom-typescript-ast-transformer-731e2b0b66e6

**Comment by adventure-yunfei** [comment]
> An idea might be to preprocess the files to replace all the `import * as AliasedName from 'moduleName'` into
> 
> ```
> import {
>   A as tmpA,
>   B as tmpB,
>   C as tmpC
> } from 'moduleName';
> declare namespace AliasedName {
>   export {
>     tmpA as A,
>     tmpB as B,
>     tmpC as C
>   }
> }
> ```

btw have you tried this manually? I've tried to support the case in namespace syntax proposal but it seems it's not valid in typescript.

**Comment by paztis** [comment]
It seems to work in 3.7

**Comment by paztis** [comment]
@adventure-yunfei, it is possible to use your AstImportInternal managment and update the dts rollup to use the namespace syntax ?
The goal of the temporary import names is to avoid conflicts with the already present imports

**Comment by adventure-yunfei** [comment]
> It seems to work in 3.7

That's awesome! Tested that only typescript@>3.7 support this syntax for namespace. (see [ts playground](https://www.typescriptlang.org/play/index.html#code/JYWwDg9gTgLgBAbzgMwhOBfFUIjgcgGdcBTAWhAgBMBXAGxPwG4AoFgYzoENDC4AhLlEQY2VEpyEk4AOy4gShMF3bSAcgGUA+hprtVvRCzgm4JAB6RYiFGkytTZy9HhJBwjA9MWrrgULgeOHFkLnp4TxZRFjkFJRV1bQAxLmA6I0cfFxtUdEjM52s3AMjo9ggZQnguOABeOE0dPQNCADp3VnLK+AAjOobtXX1FNtzOiqq4dn7GoZbWkLC6GCYgA))


> @adventure-yunfei, it is possible to use your AstImportInternal managment and update the dts rollup to use the namespace syntax ?
> The goal of the temporary import names is to avoid conflicts with the already present imports

I've created new AstEntity for local import and made it work for api-documenter. It should be able to support dts, but that will be a tough task. For temporary usage, writting a ts transformer is much easier.

**Comment by paztis** [comment]
I'm trying to do it, but hte problem is ExportAnalyzer._tryMatchImportDeclaration accept to return only one AstEntity.

Here if I apply my transformation I will have to declare X new imports + 1 AstSymbol for hte namespace.
Or I need to do this before the ExportAnalyzer job.
Any idea of the good place to do it ?

**Comment by adventure-yunfei** [comment]
Finally I got some time to continue work on this.
With previous work that analyze local namespace import as `AstImportInternal`, the only thing is to implement dts rollup generator for this type of entity. After investigating typescript@3.7 namespace syntax, I came up with cleaner dts result proposal. Unlike the proposals above, the rule is very simple while covering all cases.

**New proposal for dts with local namespace import:**

1. Declare exports of local-namespace-import entity on top level.
2. Declare local-namespace-import as `namespace`.
3. Export references (alias) inside the namespace.

With these rules, following cases are all supported inside local namespace import (and no hacky output):

- export declarations from external module
- export `default`
- re-export (regular declarations, module, or nested module)

e.g. with following inputs:
```typescript
// index.d.ts
import * as mod from './mod';
import * as mod2 from './mod2';
import { Foo } from './mod';
export {
  mod,
  mod2,  // re-export entire "mod2" module
  Foo      // re-export declaration from "mod" module
};

// mod.d.ts
import * as mod2 from './mod2';
import { ExternalDeclaration } from 'external-module';
import * as ExternalModule from 'external-module';
interface IndirectDeclaration {}  // a referenced (but not exported) declaration
export interface Foo {
    mem: IndirectDeclaration;
}
export { Foo as Foo2 };   // re-export inside namespace imported module
export default Foo;  // default export
export { mod2 };
export { ExternalDeclaration };  // export from external module
export { ExternalModule };

// mod2.d.ts
export interface Mod2Declaration {}
```

Produce dts result:
```typescript
import { ExternalDeclaration } from 'external-module';
import * as ExternalModule from 'external-module';

export declare interface Foo {
    mem: IndirectDeclaration;
}

declare interface IndirectDeclaration {}

declare namespace mod {
  export {
    Foo,
    Foo as Foo2,
    Foo as default,
    mod2_2 as mod2,
    ExternalDeclaration,
    ExternalModule,
  }
}
export { mod }

declare namespace mod2 {
  export {
    Mod2Declaration,
  }
}
export { mod2 }

declare namespace mod2_2 {
  export {
    Mod2Declaration,
  }
}

declare interface Mod2Declaration {}

export { }
```

The implementation is updated in https://github.com/adventure-yunfei/web-build-tools/pull/2. I'm still testing more cases, but have a try if needed.

**Comment by paztis** [comment]
Seems to be correct.
I just encounter one problem with this usage: #1754
But it is not related to local namespaces

I continue to test it (I've push your code on my fork to test it)

**Comment by paztis** [comment]
I encounter 1 error:
```
InternalError: Internal Error: Unsupported AstImportInternal

You have encountered a software defect. Please consider reporting the issue to the maintainers of this application.
    at Function._checkReferences (C:\Work\Workspaces\personal_workspaces\workspace_rushstack\rushstack\apps\api-extractor\lib\enhancers\ValidationEnhancer.js:155:23)
    at C:\Work\Workspaces\personal_workspaces\workspace_rushstack\rushstack\apps\api-extractor\lib\enhancers\ValidationEnhancer.js:18:44
```

Try to find the exact origin and I will post you the info

**Comment by adventure-yunfei** [comment]
> I encounter 1 error:

This only supports for api-model and dts rollup feature now.
Could you disable other features (usually "apiReport" feature) and try again?

**Comment by paztis** [comment]
I already disable everything:
```
{
  "mainEntryPointFilePath": "...",
  "bundledPackages": [...],
  "apiReport": {
    "enabled": false
  },
  "docModel": {
    "enabled": false
  },
  "dtsRollup": {
    "enabled": true,
    "untrimmedFilePath": "<projectFolder>/<unscopedPackageName>.d.ts"
  }
}
```

**Comment by paztis** [comment]
Here is the call stack: 
```
You have encountered a software defect. Please consider reporting the issue to the maintainers of this application.
    at Function._checkReferences (rushstack\apps\api-extractor\lib\enhancers\ValidationEnhancer.js:155:23)
    at rushstack\apps\api-extractor\lib\enhancers\ValidationEnhancer.js:18:44
    at AstDeclaration.forEachDeclarationRecursive (rushstack\apps\api-extractor\lib\analyzer\AstDeclaration.js:133:9)
    at AstDeclaration.forEachDeclarationRecursive (rushstack\apps\api-extractor\lib\analyzer\AstDeclaration.js:135:19)
    at AstSymbol.forEachDeclarationRecursive (rushstack\apps\api-extractor\lib\analyzer\AstSymbol.js:100:28)
    at Function.analyze (rushstack\apps\api-extractor\lib\enhancers\ValidationEnhancer.js:17:38)
    at Function.invoke (rushstack\apps\api-extractor\lib\api\Extractor.js:99:49)
```

**Comment by adventure-yunfei** [comment]
> You have encountered a software defect

I see. There're some defensive code to prevent edge cases. This edge case should be, referring a namespace-import declaration like:
```ts
import * as someModule from './local-file';
export declare class SomeClass {
    getSomething(): someModule.SomeAPI;
}
```

which have a workaround as:
```ts
import { SomeAPI } from './local-file';
export declare class SomeClass {
    getSomething(): SomeAPI;
}
```

There're still some other edge cases that won't be handled yet.  The example code is simple. Let me have a test on a real project and fix them first.

**Comment by adventure-yunfei** [comment]
@paztis, I confirmed that the error is caused by internal validation, which can be ignored here.
I've made a little change, and tested on my internal app framework package. seems it works well.
Have a re-try.

The only unsupported case I found is: using star export for **external** module inside a namespace-imported **local** module:
```ts
// index.d.ts
import * as localModule from './local-module';
export { localModule };

// mod1.d.ts
export * from 'external-module';
```

This is really a very special case, and I didn't found a way to represent that with current namespace syntax. Maybe we could leave that to future typescript support.

**Comment by zxc122333** [comment]
Thanks @adventure-yunfei, your implementation works very well for me (pixi.js). Do you have any update on this?

Since many people are waiting for this feature, is it ok for api-extractor merge the implementation of @adventure-yunfei as a first step solution? Leave an exception as current for unsupported special cases.

Or if the special cases have to be resolved, is there anything I can do to help?

**Comment by resynth1943** [comment]
Bump

**Comment by mckn** [comment]
I have opened this PR: https://github.com/microsoft/rushstack/pull/1796 please feel free to help out with suggestions on how to improve it to get it merged faster.

**Comment by octogonz** [comment]
@mckn Thank you very much for creating a PR!  I apologize that haven't been keeping up with this thread (life has been pretty crazy recently!).  I will look at your PR as soon as I get a chance. Thanks!

**Comment by resynth1943** [comment]
Sounds dope. Can't wait for this to get implemented 😛

**Comment by adventure-yunfei** [comment]
seems this feature is coming soon!

**Comment by paztis** [comment]
Hey what is the status for it ?
We were having a solution for this since March and seems still not merged on master.

Anyone works on its merge ?

**Comment by octogonz** [comment]
I was working on this, but got sandbagged by a bunch of high priority work for the Heft and Rush projects, and have not been able to give much time to API Extractor. Sorry about that... the pressure should hopefully be off by end of next week.

**Comment by Jack-Works** [comment]
Also, please support `export * as util from './utils'` syntax

**Comment by bengfarrell** [comment]
Much appreciated on the status updates from about 3 weeks ago. My team is finding that typedoc isn't up to par for our use, and checked on this project due to a Twitter thread recommending api-extractor this morning. This issue is preventing us from integrating. But we're not really in a huge rush, so we'd be happy to come back to it when this issue is fixed. Is there any sort of ETA right now? thanks!

**Comment by resynth1943** [comment]
> rush

[Hehe](https://github.com/microsoft/rushstack) ;-)

**Comment by resynth1943** [comment]
> Also, please support `export * as util from './utils'` syntax

They probably won't, unless it's been standardised?

**Comment by clar-cmp** [comment]
It has been standardised.

**Comment by octogonz** [comment]
>  My team is finding that typedoc isn't up to par for our use, and checked on this project due to a Twitter thread recommending api-extractor this morning. This issue is preventing us from integrating. But we're not really in a huge rush, so we'd be happy to come back to it when this issue is fixed. Is there any sort of ETA right now? 

@bengfarrell  Thanks for letting us know. I'll try to bump the priority of this one.

This is one of several API Extractor PRs that have been open for way-too-long because they involve architecture changes that need to be carefully considered before merging them.  In each case I need to find maybe an afternoon of time, to think through all the consequences and work out the edge cases.  It is frustrating because the PR itself does mostly work, at least for the problem cases that the submitter was concerned with.  Unfortunately there's been a firehose of development for the Rush/Heft projects lately that has consumed a lot of my attention. I apologize for that.

As an interim solution, we published the branch from PR https://github.com/microsoft/rushstack/pull/1796 as a prerelease `7.8.2-pr1796.0` -- could you try that and confirm whether it works for you or not?  If not, let me know what is broken. Thanks!

**Comment by xirzec** [comment]
> This is one of several API Extractor PRs that have been open for way-too-long because they involve architecture changes that need to be carefully considered before merging them. In each case I need to find maybe an afternoon of time, to think through all the consequences and work out the edge cases. It is frustrating because the PR itself does mostly work, at least for the problem cases that the submitter was concerned with. Unfortunately there's been a firehose of development for the Rush/Heft projects lately that has consumed a lot of my attention. I apologize for that.
> 

I have a lot of empathy for this - making architectural changes in something so complicated is very mentally demanding and anxiety inducing. I appreciate all the efforts you put in Pete!

**Comment by bengfarrell** [comment]
@octogonz Awesome, thanks! I'd have no idea that it was complicated as I don't follow the project. Thanks much for the explanation, and I'm hoping to try your suggestion in the next day or so. It's not a mission critical project, but just needs some documentation, even if a bit imperfect, so a prerelease certainly should work! I'll report back either way. Thanks again!

**Comment by bengfarrell** [comment]
Just checking back in to say that the PR release package seems to work great! Thanks for the tip!

**Comment by resynth1943** [comment]
> It has been standardised.

It's still in the [ECMAScript 2021 draft](https://tc39.es/ecma262/).

**Comment by clar-cmp** [comment]
> > It has been standardised.
> 
> It's still in the [ECMAScript 2021 **draft**](https://tc39.es/ecma262/).

It's in the [TypeScript 3.8 **release**](https://www.typescriptlang.org/docs/handbook/modules.html#export-all-as-x). :wink:

**Comment by resynth1943** [comment]
> > > It has been standardised.
> > 
> > 
> > It's still in the [ECMAScript 2021 **draft**](https://tc39.es/ecma262/).
> 
> It's in the [TypeScript 3.8 **release**](https://www.typescriptlang.org/docs/handbook/modules.html#export-all-as-x). wink

That **release** still doesn't make it a standard feature. 😉

**Comment by schmidt-sebastian** [comment]
FWIW, `7.8.2-pr1796.0` doesn't work in our repo (https://github.com/firebase/firebase-js-sdk). It seems to break since we have standalone type definition files that do not have any corresponding JS code. This is an example of where it breaks: https://github.com/firebase/firebase-js-sdk/blob/master/packages/firestore/src/api/user_data_reader.ts#L20

**Comment by transitive-bullshit** [comment]
Just ran into this on [notion-client](https://github.com/saasify-sh/notion).

@octogonz I can confirm that `7.8.2-pr1796.0` resolved my issue.

**Comment by CoreyDotCom** [comment]
Would be great to get this merged... fix is very helpful.

**Comment by clar-cmp** [comment]
> > > > It has been standardised.
> > > 
> > > 
> > > It's still in the [ECMAScript 2021 **draft**](https://tc39.es/ecma262/).
> > 
> > 
> > It's in the [TypeScript 3.8 **release**](https://www.typescriptlang.org/docs/handbook/modules.html#export-all-as-x). wink
> 
> That **release** still doesn't make it a **standard** feature. 😛

I'm actually a bit confused, then. Considering how this is a library designed for TypeScript, if it's a released TypeScript feature, isn't that what matters?

**Comment by octogonz** [comment]
> I'm actually a bit confused, then. Considering how this is a library designed for TypeScript, if it's a released TypeScript feature, isn't that what matters?

This seems like a pedantic concern. Regardless of the technical definition of "standard", if TypeScript supports a particular language feature, and using it is considered a best practice, then API Extractor will do its best to support it.

**Comment by hiranya911** [comment]
Is this issue also going to address problems like #2145 where API Extractor does not include re-exported namespace content in the API report? Fixes in 7.8.2-pr1796.0 allows me to generate a report, but it doesn't include any of the re-exported APIs.

**Comment by jasonswearingen** [comment]
```7.8.2-pr1796.0``` kind of works, but I have to disable apiReport, and the doc-model .json doesn't include re-exported content as mentioned by https://github.com/microsoft/rushstack/issues/1029#issuecomment-683347716

**Comment by jasonswearingen** [comment]
there is another, similar problem with ```export * as moduleName from "./moduleName"```  which gives the generic error message:
```
Internal Error: Unable to parse module specifier
```

**Comment by hiranya911** [comment]
Just ran into this while trying to re-export types declared in another module:

```typescript
import * as _app from './app/index';

export namespace credential {
  export type Credential = _app.Credential;
}
```

It seems, at least for cases like the above you can workaround the error by doing something like this:

```typescript
import { Credential as _Credential } from './app/index';

export namespace credential {
  export type Credential = _Credential;
}
```

**Comment by IgorMinar** [comment]
@octogonz thanks for looking into this problem. It seems that https://github.com/microsoft/rushstack/pull/1796 is a solution that is already in a pretty good shape. Could you please help us get it across the line and resolve this problem?

Fixing this issue would significantly help Angular (https://github.com/angular/angular/issues/42064) where we heavily use namespaces to reduce the size (and associated memory footprint) of generated d.ts files.

**Comment by gregjacobs** [comment]
@IgorMinar @octogonz +1 for this - we just upgraded to Angular 12 and can no longer use API extractor 😢

**Comment by octogonz** [comment]
> 
> 
> Also, please support `export * as util from './utils'` syntax

This syntax is not handled by PR #1796. It is a relatively easy addition, but in the interest of getting that PR finally merged, I'm going to postpone `export * as` to later PR.  I've created #2780 to track that.

**Comment by octogonz** [comment]
PR #1796 was released with API Extractor 7.17.0

@IgorMinar @gregjacobs Could you try it with Angular 12 and confirm whether that issue is resolved?  If not, please create a new GitHub issue with repro steps for Angular.  Thanks!

**Comment by IgorMinar** [comment]
We will. Thank you!

**Comment by alan-agius4** [comment]
Hi @octogonz, can confirm that 7.17.0 addresses the issue with Angular.

**Comment by octogonz** [comment]
> Hi @octogonz, can confirm that 7.17.0 addresses the issue with Angular.

Awesome! 🎉
